### PR TITLE
[WIP] cleanup tabs, seperate reducers

### PR DESCRIPTION
**Pull request by bomsy**
Associated Issue: #3279

### Summary of Changes

* Adds tab reducer 
* cleanup work

####Todo
* Fix lint issues
* Fix tests

### CODE REVIEW

File: src/actions/sources/select.js [hunk 1]
```diff
@@ -79,7 +84,7 @@ export function selectSourceURL(
  * @memberof actions/sources
  * @static
  */
-export function selectSource(sourceId: string, tabIndex: string = "") {
+export function selectSource(sourceId: string, tabIndex: number = undefined) {
```
Comment (wldcordeiro) [inline src/actions/sources/select.js]: No real point in an `undefined` default, that would be the case if you just left it off.

File: src/actions/tabs.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  };
+}
+
+/**
+ * @memberof actions/sources
+ * @static
+ */
+export function closeTab(id: string) {
+  return { type: "CLOSE_TAB", id };
+  /*return ({ dispatch, getState, client }: ThunkArgs) => {
```
Comment (wldcordeiro) [inline src/actions/tabs.js]: Delete these comments?

Comment (bomsy) [reply src/actions/tabs.js]: yup!

**Comment by jasonLaster** [comment]
looks good. badly needs a rebase, so i'll review after that :P

File: src/reducers/tabs.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * Gets the next tab to select when a tab closes. Heuristics:
+ * 1. if the selected tab is available, it remains selected
+ * 2. if it is gone, the next available tab to the left should be active
+ * 3. if the first tab is active and closed, select the second tab
+ *
+ * @memberof reducers/sources
+ * @static
+ */
+export function selectNewTab(state: OuterState, availableTabs: any): number {
```
Comment (bomsy) [inline src/reducers/tabs.js]: should not export this

File: src/reducers/tabs.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  return prefsTabs;
+}
+
+type OuterState = { tabs: Record<TabsState> };
+
+const getTabsState = state => state.tabs;
+
+export const getTabs = createSelector(getTabsState, tabs => tabs.tabs);
+
+export function getTabIndex(state: OuterState) {
```
Comment (bomsy) [inline src/reducers/tabs.js]: getSelectedTabIndex might be clearer

File: src/actions/tabs.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * @static
+ */
+export function closeTab(id: string) {
+  return async ({ dispatch, getState, client }: ThunkArgs) => {
+    removeDocument(id);
+
+    await dispatch({ type: "CLOSE_TAB", id });
+
+    const tab = getSelectedTab(getState());
+    const index = getTabIndex(getState());
```
Comment (bomsy) [inline src/actions/tabs.js]: Might be able to squash these two calls into one

**Comment by darkwing** [comment]
Hey @bomsy!  Anything we can do to help here?  Hope you're well!

**Comment by bomsy** [comment]
@darkwing i'm good thanks alot, just coming off a hectic week, i'm fixing and writing the tests for these atm .. its a bit slow but will soon be done

File: src/actions/sources/select.js [hunk 6]
```diff
@@ -109,7 +114,17 @@ export function selectLocation(location: Location, tabIndex: string = "") {
...
+    dispatch(
+      addTab(
+        {
+          id: src.id,
+          tooltip: getFileURL(src),
+          title: getFilename(src)
+        },
+        tabIndex
+      )
+    );
```
Comment (jasonLaster) [inline src/actions/sources/select.js]: do we need `addTab` here or could the tab reducer listen for `Select_SOURCE`? 

not sure :)

**Comment by darkwing** [comment]
Closing in favor of rebase https://github.com/devtools-html/debugger.html/pull/5672

**Pull request closed without merging**